### PR TITLE
Add drop table support

### PR DESCRIPTION
**Pull request by Lujie1996**
## Overview

Description:
Added a CorfuTable dropTable API which removesa table
from TableRegistry and clear its record in the sequencer's
in-memory maps. The effects of invoking this API is
persisted even after server restarts.

Why should this be merged: 
Avoid RegistryTable leak, and its consequence of TrimmedException when reopening a trimmed table.

Related issue(s) (if applicable): #3251 

## Checklist (Definition of Done):

- [ ] There are no TODOs left in the code
- [ ] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [ ] Change is covered by automated tests
- [ ] Public API has Javadoc

### CODE REVIEW

**Comment by hisundar** [comment]
Is the idea to do a soft delete and have the checkpointer do the actual delete of the stream?

**Comment by Lujie1996** [comment]
> Is the idea to do a soft delete and have the checkpointer do the actual delete of the stream?

Hi Sundar. This PR is currently WIP. The purpose of this PR is to support a drop table API that clears all the stream and table information from the RegistryTable and Sequencer server in-memory states.

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java [hunk 1]
```diff
@@ -784,13 +842,32 @@ private void handleAllocation(RequestMsg req, ChannelHandlerContext ctx, IServer
...
+        }
+
         // Note: we reuse the request header as the ignore_cluster_id and
         // ignore_epoch fields are the same in both cases.
         ResponseMsg response = getResponseMsg(
                 getHeaderMsg(req.getHeader()), getTokenResponseMsg(newToken, backPointerMap.build()));
         r.sendResponse(response, ctx);
     }
 
+    private Set<UUID> getIllegalStreamsFromRequest(TxResolutionInfo txResolutionInfo) {
```
Comment (PavelZaytsev) [inline infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java:857]: What does it mean for a stream to be illegal?

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/TokenType.java [hunk 2]
```diff
@@ -25,7 +25,9 @@ public enum TokenType {
 
     // Sent when a transaction aborts because it has an old version (i.e. older than
     // the trim mark). This is to detect slow transactions
-    TX_ABORT_SEQ_TRIM((byte) 4);
+    TX_ABORT_SEQ_TRIM((byte) 4),
+
+    TX_ABORT_ILLEGAL_STREAM((byte) 5);
```
Comment (PavelZaytsev) [inline runtime/src/main/java/org/corfudb/protocols/wireprotocol/TokenType.java:30]: Plz provide description

**Review by PavelZaytsev — APPROVED** [review]
lgtm

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java [hunk 3]
```diff
@@ -588,12 +605,43 @@ public void resetServer(@Nonnull RequestMsg req,
                 " sequencerEpoch = {}", globalLogTail, streamTailToGlobalTailMap.size(),
                 streamTailToGlobalTailMap, sequencerEpoch);
 
+        if (serverContext.isDeleteInactiveStreamsOnSequencerReset()) {
+            CompletableFuture<Set<UUID>> getInactiveStreamsFuture = CompletableFuture.supplyAsync(() ->
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java:609]: Wouldn't this introduce a race condition where the sequencer is bootstrapped (allow for other clients to be serviced) but the async task is still performing a clean up on the metatdata?

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java [hunk 4]
```diff
@@ -588,12 +605,43 @@ public void resetServer(@Nonnull RequestMsg req,
...
+                            getInactiveStreams(Collections.unmodifiableSet(streamsAddressMap.keySet())),
+                            corfuRuntimeExecutor)
+                    .exceptionally(ex -> {
+                        log.error("Failed to get inactive streams.", ex);
+                        return Collections.emptySet();});
+            getInactiveStreamsFuture.thenRunAsync(() -> deleteStreams(getInactiveStreamsFuture.join()),
+                            executor)
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java:616]: I would suggest simplifying the threading model. Instead of executing one stage of the async flow on one executor and the other stage on another executor, I think its cleaner to just do it in a synchronous manner on on executor

Comment (Lujie1996) [reply infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java:616]: That introduces dead lock, doesn't it? Getting the inactive streams needs corfu runtime and table read on RegistryTable. That's why I unblock sequencer main thread here and put the table read on a different thread.

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java (hunk 3 above)
Comment (Lujie1996) [reply infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java:609]: Yes. There will be a very short window after the sequencer bootstrap. But there will not be any correctness issue since a dropped table has a CLEAR update. Today there is no way to prevent the sequencer from serving requests once bootstrap is done. We will need to have two-phase sequencer bootstrap and client-aware sequencer request control to achieve that.

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java [hunk 5]
```diff
@@ -865,6 +942,50 @@ private Map<UUID, StreamAddressSpace> getStreamsAddressesMap(
...
+    }
+
+    private CorfuRuntime getNewCorfuRuntime() {
+        final CorfuRuntime.CorfuRuntimeParameters params =
+                serverContext.getManagementRuntimeParameters();
+        params.setSystemDownHandlerTriggerLimit(SYSTEM_DOWN_HANDLER_TRIGGER_LIMIT);
+        final CorfuRuntime runtime = CorfuRuntime.fromParameters(params);
+        final Layout managementLayout = serverContext.copyManagementLayout();
+        // Runtime can be set up either using the layout or the bootstrapEndpoint address.
+        if (managementLayout != null) {
```
[the remark this answers is not in the document]
Comment (Lujie1996) [reply infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java:980]: This code is actually taken from ManagementServer.

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/TokenResponse.java [hunk 6]
```diff
@@ -20,6 +21,8 @@
     public static byte[] NO_CONFLICT_KEY = new byte[]{0};
     public static UUID NO_CONFLICT_STREAM = new UUID(0, 0);
 
+    public static Set<UUID> NO_ILLEGAL_STREAM = Collections.emptySet();
```
[the remark this answers is not in the document]
Comment (Lujie1996) [reply runtime/src/main/java/org/corfudb/protocols/wireprotocol/TokenResponse.java:24]: Done

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/TokenResponse.java [hunk 7]
```diff
@@ -33,6 +36,7 @@
         this.token = token;
         this.backpointerMap = backpointerMap;
         this.streamTails = Collections.emptyMap();
+        this.illegalStreams = Collections.emptySet();
```
[the remark this answers is not in the document]
Comment (Lujie1996) [reply runtime/src/main/java/org/corfudb/protocols/wireprotocol/TokenResponse.java:39]: Done.

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java (hunk 1 above)
Comment (Lujie1996) [reply infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java:857]: Added a comment in the code. A stream in transaction createSet is considered illegal if the transaction attempts to create it while Sequencer has already allocated token for it. This is used to detect the 'reopening a trimmed table' scenario.

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java (hunk 4 above)
Comment (Maithem) [reply infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java:616]: Sorry, I think my comment wasn't clear, by

> synchronous manner on on executor

I meant why not just execute the logic on `corfuRuntimeExecutor`. Essentially, I don't understand why the workflow needs to be split across two executors.

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/TokenType.java (hunk 2 above)
Comment (Lujie1996) [reply runtime/src/main/java/org/corfudb/protocols/wireprotocol/TokenType.java:30]: Added the following

    // Sent when a transaction aborts because it attempts to create a stream that
    // already exists in sequencer. This is to detect when a client tries to reopen
    // an already trimmed stream.

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java (hunk 4 above)
Comment (Lujie1996) [reply infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java:616]: We know that this logic shouldn't block the main sequencer thread (deadlock), and the sequencer will serve other requests while the corfuRuntimeExecutor does the deleteStream. In that case, we need to synchronize between the two sequencer threads, which is even more complicated. That's why I think keeping the main sequencer thread the ONLY thread that modifies the maps is most straightforward.

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java [hunk 8]
```diff
@@ -865,6 +955,64 @@ private Map<UUID, StreamAddressSpace> getStreamsAddressesMap(
...
+            .collect(Collectors.toList());
+        List<UUID> checkpointStreams = realStreams.stream()
+            .map(CorfuRuntime::getCheckpointStreamIdFromId)
+            .collect(Collectors.toList());
+
+        List<UUID> droppedStreams = new ArrayList<>();
+        droppedStreams.addAll(realStreams);
+        droppedStreams.addAll(checkpointStreams);
+
+        corfuRuntime.shutdown();
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java:989]: This looks like a resource leak.

What happens if there is an exception thrown before `corfuRuntime.shutdown()` ? (Memory, Threads and file descriptors will not be freed)

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (Maithem) [reply infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java:1000]: I suggest changing the if-condition to an assertion/precondition `if (managementLayout != null)`

The code **expects** the management layout to exist, otherwise the runtime doesn't know which servers to connect to.

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java [hunk 9]
```diff
@@ -865,6 +955,64 @@ private Map<UUID, StreamAddressSpace> getStreamsAddressesMap(
...
+        if (managementLayout != null) {
+            managementLayout.getLayoutServers().forEach(runtime::addLayoutServer);
+        }
+        runtime.connect();
+        log.info("getCorfuRuntime: Corfu Runtime connected successfully");
+        params.setSystemDownHandler(runtimeSystemDownHandler);
+        return runtime;
+    }
+
+    private boolean isRegistryTableOpened() {
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java:1009]: I think the nomenclature can be improved further for clarity. Consider using initialized instead of Opened?

File: infrastructure/src/test/java/org/corfudb/infrastructure/management/FailureDetectorTest.java [hunk 10]
```diff
@@ -46,7 +46,7 @@ public void testPollRound() {
         );
         Duration time = Duration.ofMillis(System.currentTimeMillis() - start);
         assertThat(time).isGreaterThan(Duration.ofMillis(450));
-        assertThat(time).isLessThan(Duration.ofSeconds(2));
+        assertThat(time).isLessThan(Duration.ofSeconds(20));
```
Comment (Maithem) [inline infrastructure/src/test/java/org/corfudb/infrastructure/management/FailureDetectorTest.java:49]: Interesting, my patch is also failing on this test.

I feel like this number should be lower and perhaps fixed in a separate patch.

File: runtime/proto/service/sequencer.proto [hunk 11]
```diff
@@ -59,6 +59,8 @@ message TokenResponseMsg {
     // Sent when a transaction aborts because it has an old version (i.e. older than
     // the trim mark). This is to detect slow transactions
     TX_ABORT_SEQ_TRIM = 4;
+
+    TX_ABORT_ILLEGAL_STREAM = 5;
```
Comment (Maithem) [inline runtime/proto/service/sequencer.proto:63]: Invalid stream?

File: runtime/src/main/java/org/corfudb/runtime/view/TableRegistry.java [hunk 12]
```diff
@@ -672,11 +706,84 @@ public void freeTableData(String namespace, String tableName) {
...
+            try {
+                this.runtime.getObjectsView().TXBuild()
+                        .type(TransactionType.WRITE_AFTER_WRITE)
+                        .build()
+                        .begin();
+
+                // Clear table data
+                UUID streamId = CorfuRuntime.getStreamID(getFullyQualifiedTableName(namespace, tableName));
+                SMREntry entry = new SMREntry("clear", new Array[0], Serializers.PRIMITIVE);
+                TransactionalContext.getCurrentContext().logUpdate(streamId, entry);
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/view/TableRegistry.java:752]: Do you write a raw SMR entry to prevent opening the table here?

Isn't clear a pure mutator and therefore won't sync the table?

File: runtime/src/main/java/org/corfudb/runtime/view/TableRegistry.java [hunk 13]
```diff
@@ -672,11 +706,84 @@ public void freeTableData(String namespace, String tableName) {
...
+                            TableMetadata.newBuilder().mergeFrom(oldRecord.getMetadata()).setIsDropped(true).build());
+                    this.registryTable.insert(tableNameKey, newRecord);
+                }
+
+                this.runtime.getObjectsView().TXEnd();
+
+                log.info("Successfully cleared and unregister table {}${}", namespace, tableName);
+                break;
+            } catch (TransactionAbortedException txAbort) {
+                if (numRetries <= 0) {
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/view/TableRegistry.java:774]: This code assumes that side-effect of `addToDeleteSet` and the transaction commit phase to be atomic, but it isn't.

For example, the sequencer can delete the streams, but the write can fail leaving the system in an inconsistent state.

**Review by Maithem — COMMENTED** [review]
This patch makes some undocumented assumptions that I think should be articulated. First, the sequencer should have enough memory to open the registry table, right now almost all of the sequencer's data structures are fixed size (except for the request queue). Second, the pruning of dropped streams is best-effort, and if it fails, it can leave the system in an inconsistent state. For example, what would happen if the sequencer was bootstrapped, but the pruning phase never completes?


Furthermore, the added clearSet and deleteSet in `TxResolutionInfo` introduce nondeterministic behavior to the sequencer; It's not clear that this non-atomic two phase workflow to drop a stream is correct.


I think its worthwhile to spend some more time to try to refine the implementation.

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java (hunk 9 above)
Comment (Lujie1996) [reply infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java:1009]: Done.

File: infrastructure/src/test/java/org/corfudb/infrastructure/management/FailureDetectorTest.java (hunk 10 above)
Comment (Lujie1996) [reply infrastructure/src/test/java/org/corfudb/infrastructure/management/FailureDetectorTest.java:49]: Reverted back to 2. 2 is too short and it always fails for me locally.

File: runtime/proto/service/sequencer.proto (hunk 11 above)
Comment (Lujie1996) [reply runtime/proto/service/sequencer.proto:63]: Done.

File: runtime/src/main/java/org/corfudb/runtime/view/TableRegistry.java [hunk 14]
```diff
@@ -672,11 +706,84 @@ public void freeTableData(String namespace, String tableName) {
...
+
+                CorfuRecord<TableDescriptors, TableMetadata> oldRecord =
+                        this.registryTable.get(tableNameKey);
+                if (oldRecord == null || oldRecord.getPayload() == null) {
+                    log.warn("The table to unregister {} does not exist in RegistryTable.", tableNameKey);
+                } else {
+                    CorfuRecord<TableDescriptors, TableMetadata> newRecord = new CorfuRecord<>(
+                            oldRecord.getPayload(),
+                            TableMetadata.newBuilder().mergeFrom(oldRecord.getMetadata()).setIsDropped(true).build());
+                    this.registryTable.insert(tableNameKey, newRecord);
```
Comment (hisundar) [inline runtime/src/main/java/org/corfudb/runtime/view/TableRegistry.java:766]: Maybe I missed something but when is the entry removed from the registryTable if we are attempting a soft delete here?
Also how does it work with openTable() which seems to be checking for absence of the entry itself as a pre-condition for dropping before open?

**Review by hisundar — COMMENTED** [review]
Just wondering is it necessary to have a new sequencer protocol.
Won't something like this suffice
1.  In dropTable() - remove the entry from the registryTable and return
2. When a new table is being created, we do a sequencer query and if we find that the stream has an address lower than trim mark (TRIMMED) or a tail, we append the clear SMR entry and reset the stream address space so it looks like a brand new table.

Basically we only do something special if the caller is attempting to open a table and the sequencer has some stale data for that table from a previous drop right?

File: runtime/src/main/java/org/corfudb/runtime/view/TableRegistry.java (hunk 12 above)
Comment (Lujie1996) [reply runtime/src/main/java/org/corfudb/runtime/view/TableRegistry.java:752]: Yes. Not opening the table makes this API more versatile (for example, dropping a corrupted table that can't be opened).

**Comment by Lujie1996** [comment]
> Just wondering is it necessary to have a new sequencer operation that operates on table registry protocol. Won't something like this suffice
> 
> 1. In dropTable() - remove the entry from the registryTable and return
> 2. When a new table is being created, we do a sequencer query and if we find that the stream has an address lower than trim mark (TRIMMED) or a tail, we append the clear SMR entry and reset the stream address space so it looks like a brand new table.
> 
> Basically we only do something special if the caller is attempting to open a table and the sequencer has some stale data for that table from a previous drop right?

In the case where the table has been trimmed, you will still need to clean up the sequencer maps in order to append the CLEAR SMR entry and for any client to read the table after reopen. I think overall what you proposed is similar to the existing patch. The difference is at when you actually clean up the sequencer maps.

File: runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java [hunk 15]
```diff
@@ -680,6 +681,10 @@ public void prefixTrim(final Token address, boolean retry) {
         }
     }
 
+    public void resetStream(List<UUID> streams) {
```
[the remark this answers is not in the document]
Comment (Lujie1996) [reply runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java]: There is no such method in the latest patch.

File: runtime/src/main/java/org/corfudb/runtime/view/TableRegistry.java (hunk 14 above)
Comment (Lujie1996) [reply runtime/src/main/java/org/corfudb/runtime/view/TableRegistry.java:766]: 1. I will work on that in a later PR. https://github.com/CorfuDB/CorfuDB/pull/3540#issuecomment-1487408937 I'm thinking to record the timestamp at which the table is marked as dropped, and based on this info let compactor clean up the entry.
2. A normal openTable() called on a table for the first time will pass the sequencer check. Only a table not existing in RegistryTable but exist in sequencer maps will be invalidated by sequencer. https://github.com/CorfuDB/CorfuDB/pull/3540/files/d45db7870d54786b944a43f54575eb22ae1cfa0f#diff-3ef8175cea450001f0ec6ce1f7f00df50a3626212a0c463cffcb464b94e95c1bR303

**Comment by Lujie1996** [comment]
> This patch makes some undocumented assumptions that I think should be articulated. First, the sequencer should have enough memory to open the registry table, right now almost all of the sequencer's data structures are fixed size (except for the request queue). Second, the pruning of dropped streams is best-effort, and if it fails, it can leave the system in an inconsistent state. For example, what would happen if the sequencer was bootstrapped, but the pruning phase never completes?
> 
> Furthermore, the added clearSet and deleteSet in `TxResolutionInfo` introduce nondeterministic behavior to the sequencer; It's not clear that this non-atomic two phase workflow to drop a stream is correct.
> 
> I think its worthwhile to spend some more time to try to refine the implementation.

The RegistryTable is not very large and the memory is released right after reading the table. I agree that when the scale goes higher the memory can be a problem.

I added `deletedStreamsSet` in Sequencer so that a token allocation request will check against this set to ensure no client is writing to a dropped table. The `deletedStreamsSet` is persistent across restarts - it's built by reading the RegistryTable at sequencer bootstrap.

Here is the semantics for the dropTable API in both normal and failure cases.

**_Normal_**
After dropTable() is completed successfully,
1. Sequencer reset does NOT reload the address space of a dropped tables.
2. Existing transactions holding in-memory snapshots are still be able to access the table via the local copy. 
3. Transactions in existing client JVMs (started before dropTable) will still be able to read a dropped table, but it's limited to the latest cached snapshot. After a compaction cycle, all cached snapshots will be reset by runtime GC. Writes to a dropped table will be rejected and receive TAE.
4. If a table is dropped and reopened, other clients might not observe this change at all and still writes to the reopened table, although the data has been reset. <= this can be an issue due to the fact that we don't differentiate tables across reopen.

**_Failure_**
Due to the non-atomic two phase clean up, the dropTable() might be partially executed, and there can be two possible failure cases as following,
1. The Sequencer maps are cleaned up but the RegistryTable is not updated. 
2. The Sequencer maps are not cleaned up but the RegistryTable is updated.

From the implementation, if Sequencer map cleanup fails, the dropTable transaction will be aborted and the RegistryTable will not be updated. Hence, the 1st case is the only possibility. In the 1st failure, after dropTable() is partially executed,

1. If sequencer is reset, the table will be loaded back again. It will behave like the table was not dropped at all.
2. If sequencer is not reset, writes to the table will get aborted and reads will reflect the last cached snapshot (same as Normal case).

I feel like there is no good way to handle this failure case. The best-effort, including 9 retries of dropping table (idempotent), if all fail how can we ensure any redo will work? I'm out of ideas...

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStoreShim.java [hunk 16]
```diff
@@ -97,7 +97,7 @@ Table<Queue.CorfuGuidMsg, V, Queue.CorfuQueueMetadataMsg> openQueue(@Nonnull Str
      * @param tableName Table name.
      */
     public void deleteTable(String namespace, String tableName) {
```
Comment (hisundar) [inline runtime/src/main/java/org/corfudb/runtime/collections/CorfuStoreShim.java:99]: I think anyone using deleteTable() is probably looking for dropTable() so we might as well remove this method rather than just make it clear the table in my opinion. It will reduce our API footprint

**Review by hisundar — COMMENTED** [review]
Thanks George, the delete of sequencer and log unit looks good, but I did not see the change to handle a dropped & trimmed table on open?
Did I miss anything?

File: runtime/src/main/java/org/corfudb/runtime/view/TableRegistry.java [hunk 17]
```diff
@@ -673,11 +680,90 @@ public void freeTableData(String namespace, String tableName) {
...
+    public void dropTable(String namespace, String tableName) {
+        log.info("Invoking dropTable {}${}", namespace, tableName);
+
+        UUID streamId = CorfuRuntime.getStreamID(getFullyQualifiedTableName(namespace, tableName));
+        UUID cpStreamId = CorfuRuntime.getCheckpointStreamIdFromId(streamId);
+        List<UUID> streamsToDelete = Arrays.asList(streamId, cpStreamId);
+
+        // Clear the stream address space from Sequencer server cache and Log unit cache.
+        // This step is needed to support reopening a dropped table.
+        runtime.getAddressSpaceView().deleteStreamFromServerCache(streamsToDelete);
```
[the remark this answers is not in the document]
Comment (Lujie1996) [reply runtime/src/main/java/org/corfudb/runtime/view/TableRegistry.java:705]: Handling reopening a trimmed table is outside the scope of this PR. I removed the auto-detection and drop table logic in registerTable. This PR solves the problem where tables need to be dropped during data migration.

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStoreShim.java (hunk 16 above)
Comment (Lujie1996) [reply runtime/src/main/java/org/corfudb/runtime/collections/CorfuStoreShim.java:99]: I think it would be meaningful to keep the clearTable because it sounds like a very common API. And just like any SQL database, usually both clear and drop operations are supported. Whether clients really want clear or drop in their code requires a closer look into each usage, so won't it be better to make the changes at each usage?

Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/collections/CorfuStoreShim.java]: sure, sounds reasonable - we can keep it then.

**Review by hisundar — CHANGES_REQUESTED** [review]
Hi George, providing two separate api with a implicit compaction assumption under the hood is very likely going to be very error prone and does not handle the case where a table is `unregisterTable()`, compacted, trimmed but then due to a failover it never got to `dropTrimmedTable()` - in this case we will hit a trimmed exception again because of the new api

Why can't we just provide one new api - 
`dropTable()`:
1. tries to deleteStream from sequencer & LU and on failure just logs error and returns (cluster could be in mixed mode and we should not modify state)
2. if step 1 is successful at deleteStreams() then delete entry from table registry
3. log append a CLEAR SMR bypassing the object layer
-------------------
Let's talk about a few existing scenarios to support:
Case 1: All the nodes are upgraded and running new version, and there is no restart of corfu server: (Happy Case) - there will be no issue when reopening the deleted table since its stream will be reset from scratch.

Case 2: Mixed Mode: command should fail at step 1 and not modify state since we cannot support dropping tables in mixed mode and there is no way for us to enforce that this api does not get invoked in mixed mode - we should just not cause data loss *if possible*.

Case 3: After dropTable executes, all the nodes go down and come back up with full in-memory state for the dropped table and compactor runs and trims the table!

Now re-opening the table and reading it will hit a TrimmedException!
To handle this case : on `openTable()` we just check if the new table already has a trimmed address space, if so we force append a start and end checkpoint so the stream is just automatically reset even if we are a mixed mode cluster .

So we just need 1 new api `dropTable()` and 1 new behavior for openTable() of a trimmed stream. That should handle all the cases right?
